gpjax/kernels/non_euclidean/categorical.py

```diff
@@ -0,0 +1,126 @@
+
+    It stores a standard dev for each input value (i.e. the diagonal of the gram), and a lower cholesky factor for correlations.
+    It returns the corresponding values from an the gram matrix when called.
+
+    Args:
+        sdev (Float[Array, "N"]): The standard deviation parameters, one for each input space value.
+        cholesky_lower (Float[Array, "N*(N-1)//2 N"]): The parameters for the Cholesky factor of the gram matrix.
+        inspace_vals (list): The values in the input space this DictKernel works for. Stored for order reference, making clear the indices used for each input space value.
+
+    Raises:
+        ValueError: If the number of diagonal variance parameters does not match the number of input space values.
+    """
+
+    sdev: Float[Array, " N"] = param_field(jnp.ones((2,)), bijector=tfb.Softplus())
+    cholesky_lower: Float[Array, "N N"] = param_field(
+        jnp.eye(2), bijector=tfb.CorrelationCholesky()
+    )
+    inspace_vals: list = static_field(None)
+    name: str = "Dictionary Kernel"
+    input_1hot: bool = static_field(False)
+
+    def __post_init__(self):
+        if self.inspace_vals is not None and len(self.inspace_vals) != len(self.sdev):
+            raise ValueError(
+                f"The number of sdev parameters ({len(self.sdev)}) has to match the number of input space values ({len(self.inspace_vals)}), unless inspace_vals is None."
+            )
+
+    @property
+    def explicit_gram(self):
+        L = self.sdev.reshape(-1, 1) * self.cholesky_lower
+        return L @ L.T
+
+    def __call__(  # TODO not consistent with general kernel interface
+        self,
+        x: ScalarInt,
+        y: ScalarInt,
+    ):
+        r"""Compute the (co)variance between a pair of dictionary indices.
+
+        Args:
+            x (ScalarInt): The index of the first dictionary entry.
+            y (ScalarInt): The index of the second dictionary entry.
+
+        Returns
+        -------
+            ScalarFloat: The value of $k(v_i, v_j)$.
+        """
+        try:
+            x = x.squeeze()
+            y = y.squeeze()
+        except AttributeError:
+            pass
+        if self.input_1hot:
+            return self.explicit_gram[jnp.outer(x, y) == 1.0]
+        else:
+            return self.explicit_gram[x, y]
+
+    @classmethod
+    def num_cholesky_lower_params(cls, num_inspace_vals: ScalarInt) -> ScalarInt:
+        """Compute the number of parameters required to store the lower triangular Cholesky factor of the gram matrix.
+
+        Args:
+            num_inspace_vals (ScalarInt): The number of values in the input space.
+
+        Returns:
+            ScalarInt: The number of parameters required to store the lower triangle of the Cholesky factor of the gram matrix.
+        """
+        return num_inspace_vals * (num_inspace_vals - 1) // 2
+
+    @classmethod
+    def gram_to_sdev_cholesky_lower(cls, gram: Float[Array, "N N"]) -> CatKernelParams:
```

Comment:
Doesn't need to be a class method since `cls` is not used. Can be a static method.